packages/core/src/crawlers/context_pipeline.ts

```diff
@@ -0,0 +1,170 @@
+import type { Awaitable } from '@crawlee/types';
+
+import {
+    ContextPipelineCleanupError,
+    ContextPipelineInitializationError,
+    ContextPipelineInterruptedError,
+    RequestHandlerError,
+    SessionError,
+} from '../errors.js';
+
+/**
+ * Represents a middleware step in the context pipeline.
+ *
+ * @template TCrawlingContext - The input context type for this middleware
+ * @template TCrawlingContextExtension - The enhanced output context type
+ */
+export interface ContextMiddleware<TCrawlingContext extends {}, TCrawlingContextExtension extends {}> {
```

Comment:
do we really need the `extends {}` here? 

(`{}` doesn't represent an object on type level)